spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BinderFactoryConfiguration.java

```diff
@@ -0,0 +1,131 @@
+public class BinderFactoryConfiguration {
+
+	@Bean
+	@ConditionalOnMissingBean(BinderFactory.class)
+	public BinderFactory binderFactory(BinderTypeRegistry binderTypeRegistry,
+									   ChannelBindingServiceProperties channelBindingServiceProperties) {
+
+		Map<String, BinderConfiguration> binderConfigurations = new HashMap<>();
+		if (!CollectionUtils.isEmpty(channelBindingServiceProperties.getBinders())) {
+			for (Map.Entry<String, BinderProperties> binderEntry :
+					channelBindingServiceProperties.getBinders().entrySet()) {
+				BinderProperties binderProperties = binderEntry.getValue();
+				if (binderTypeRegistry.get(binderEntry.getKey()) != null) {
+					binderConfigurations.put(binderEntry.getKey(),
+							new BinderConfiguration(binderTypeRegistry.get(binderEntry.getKey()),
+									binderProperties.getProperties()));
+				}
+				else {
+					Assert.hasText(binderProperties.getType(), "No 'type' property present for custom " +
+							"binder " + binderEntry.getKey());
+					binderConfigurations.put(binderEntry.getKey(),
+							new BinderConfiguration(binderTypeRegistry.get(binderProperties.getType()),
+									binderProperties.getProperties()));
+				}
+			}
+		}
+		else {
+			for (Map.Entry<String, BinderType> entry : binderTypeRegistry.getAll().entrySet()) {
+				binderConfigurations.put(entry.getKey(),
+						new BinderConfiguration(entry.getValue(), new Properties()));
+			}
+		}
+		DefaultBinderFactory binderFactory = new DefaultBinderFactory<>(binderConfigurations);
+		binderFactory.setDefaultBinder(channelBindingServiceProperties.getDefaultBinder());
+		return binderFactory;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(BinderTypeRegistry.class)
+	public BinderTypeRegistry binderTypeRegistry(ConfigurableApplicationContext configurableApplicationContext) {
+		Map<String, BinderType> binderTypes = new HashMap<>();
+		ClassLoader classLoader = configurableApplicationContext.getClassLoader();
+		if (classLoader == null) {
+			classLoader = ChannelBindingAutoConfiguration.class.getClassLoader();
+		}
+		try {
+			Enumeration<URL> resources = classLoader.getResources("META-INF/spring.binders");
+			while (resources.hasMoreElements()) {
+				URL url = resources.nextElement();
+				UrlResource resource = new UrlResource(url);
+				for (BinderType binderType : parseBinderConfigurations(classLoader, resource)) {
+					binderTypes.put(binderType.getDefaultName(), binderType);
+				}
+			}
+		}
+		catch (IOException | ClassNotFoundException e) {
+			throw new BeanCreationException("Cannot create binder factory:", e);
+		}
+		return new DefaultBinderTypeRegistry(binderTypes);
+	}
+
+
+	static Collection<BinderType> parseBinderConfigurations(ClassLoader classLoader, Resource resource)
+			throws IOException, ClassNotFoundException {
+		Properties properties = PropertiesLoaderUtils.loadProperties(resource);
+		Collection<BinderType> parsedBinderConfigurations = new ArrayList<>();
+		for (Map.Entry<?, ?> entry : properties.entrySet()) {
```

Comment:
We can do that - it's a little bit less efficient, but it doesn't matter, if you feel it reads better.